### PR TITLE
[#19] Use form to split device files into multiple folders

### PR DESCRIPTION
**Pull request by sjroe**
Implements #19

### CODE REVIEW

**Comment by adamgreig** [comment]
Thanks for implementing this! I'll try and check it over this evening but two quick notes first:

* Do you delete the various extern crate and config lines that the first part of the sed command used to delete, or do they stay in the new mod.rs? (We delete lines 1-6 and 10 of the svd2rust lib.rs).
* We can't use `sed -i` (see #25) so prefer just running sed and outputting to a new file and deleting the old one... (yes this is pretty annoying, blame OS X!)

**Comment by sjroe** [comment]
Apologies, I mistakenly removed the line removal when I was experimenting with not running `rustfmt` before making running the `sed` line.  This has been reworked.

I wasn't aware of OS X's version of `sed` didn't support -i but should have assumed that you weren't using it for a good reason.

**Comment by adamgreig** [comment]
I've tested this out and it seems to work fine (thanks for fixing it up!) but I'm not sure whether to merge. Using form increases generation time about 4x on my desktop (12s -> 39s), without changing crate compilation or documentation time noticeably. Currently the auto-generated files are not uploaded to GitHub (since there are many hundreds of megabytes of them in total), so the nicer underlying files would only be present on the crates.io archive. I don't think it helps us generate the crate documentation (#3), since they still require shocking quantities of RAM and disk.

Basically I'm not sure what the advantage would be to using Form in this case, but I'm still open to merging it if there are good reasons. Maybe you or @gbip would have some ideas?

**Comment by phil-opp** [comment]
Another advantage of using form is that you can look at the source more easily (without the need to open a file with 400k+ lines). This is especially important when using the `src` link in rustdoc, since loading this amount of lines in the browser stresses my laptop considerably. Also, it seems like rustdoc/firefox is not able not handle anchors to lines >50k (jumps to and highlights a wrong line).

**Comment by adamgreig** [comment]
That's a good point. I don't remember the memory usage being any lower but I'll check.

If rustdoc can't build them anyway then perhaps an option is to just have using Form be an option when making which is disabled by default -- when building locally (whether to generate your own documentation or just for browsing the source) you could have it enabled, but otherwise the default (and quicker) option is to skip that step.

**Comment by adamgreig** [comment]
Memory usage isn't any better sadly. The STM32F4 crate took over 12G to build all features in 9 minutes on my desktop, but I think something might have got oom killed as it finished pretty much right as it maxed out my RAM. Even viewing the generated documentation is a bit intense.

**Comment by phil-opp** [comment]
> Memory usage isn't any better sadly.

That's a pity, but thanks for checking!

> If docs.rs can't build them anyway then perhaps an option is to just have using Form be an option when making which is disabled by default -- when building locally (whether to generate your own documentation or just for browsing the source) you could have it enabled, but otherwise the default (and quicker) option is to skip that step.

The version on crates.io is pre-generated, right? So we can't just enable a cargo feature in the crate, but would need to clone and re-generate the project. Wouldn't it make more sense to just use form by default and accept the longer generation times, because generation is only done seldomly (on new releases) while `cargo doc` is used by most users (especially if there is no docs.rs documentation)?

**Comment by adamgreig** [comment]
Good point - so maybe the versions uploaded to Cargo are built using Form, but have an option to not use it when building locally for test/development where it meaningfully slows down that cycle.

**Comment by phil-opp** [comment]
Sounds good to me!

**Comment by adamgreig** [comment]
I modified this to have an optional target 'form' in the Makefile which can be called to run Form over all the generated sources, but isn't run by default. I can use it before building releases, but otherwise it's very slow and the huge number of files slows down git operations etc a lot too.
